tesseract_motion_planners/ompl/src/profile/ompl_composite_profile_rvss.cpp

```diff
@@ -0,0 +1,360 @@
+
+  if (mi.tcp_frame.empty())
+    throw std::runtime_error("OMPL: tcp_frame is empty!");
+
+  if (mi.working_frame.empty())
+    throw std::runtime_error("OMPL: working_frame is empty!");
+
+  // Get the kinematics group for solving IK
+  tesseract_kinematics::KinematicGroup::ConstPtr manip =
+      env->getKinematicGroup(mi.manipulator, mi.manipulator_ik_solver);
+
+  /** Solve IK
+   *  @todo Need to add Descartes pose sample to ompl profile
+   */
+  Eigen::Isometry3d tcp_offset = env->findTCPOffset(mi);
+  Eigen::Isometry3d tcp_frame_cwp = cartesian_waypoint * tcp_offset.inverse();
+  tesseract_kinematics::KinGroupIKInput ik_input(tcp_frame_cwp, mi.working_frame, mi.tcp_frame);
+  const tesseract_kinematics::IKSolutions joint_solutions =
+      manip->calcInvKin({ ik_input }, Eigen::VectorXd::Zero(manip->numJoints()));
+
+  // Check collision
+  std::vector<tesseract_collision::ContactResultMap> contact_map_vec(static_cast<std::size_t>(joint_solutions.size()));
+  tesseract_common::KinematicLimits limits = manip->getLimits();
+  tesseract_kinematics::IKSolutions valid_solutions;
+  valid_solutions.reserve(joint_solutions.size());
+  for (std::size_t i = 0; i < joint_solutions.size(); ++i)
+  {
+    Eigen::VectorXd solution = joint_solutions[i];
+
+    // Check limits
+    if (tesseract_common::satisfiesPositionLimits(solution, limits.joint_limits))
+    {
+      tesseract_common::enforcePositionLimits(solution, limits.joint_limits);
+    }
+    else
+    {
+      CONSOLE_BRIDGE_logDebug("In OMPLDefaultPlanProfile: Goal state has invalid bounds");
+      continue;
+    }
+
+    // Get discrete contact manager for testing provided start and end position
+    // This is required because collision checking happens in motion validators now
+    // instead of the isValid function to avoid unnecessary collision checks.
+    contact_map_vec[i] = checkCollision(solution, manip, env->getDiscreteContactManager());
+    if (contact_map_vec[i].empty())
+    {
+      valid_solutions.push_back(solution);
+
+      auto redundant_solutions = tesseract_kinematics::getRedundantSolutions<double>(
+          solution, limits.joint_limits, manip->getRedundancyCapableJointIndices());
+
+      valid_solutions.insert(valid_solutions.end(), redundant_solutions.begin(), redundant_solutions.end());
+    }
+  }
+
+  if (valid_solutions.empty())
+  {
+    for (std::size_t i = 0; i < contact_map_vec.size(); i++)
+      for (const auto& contact_vec : contact_map_vec[i])
+        for (const auto& contact : contact_vec.second)
+          CONSOLE_BRIDGE_logError(("Solution: " + std::to_string(i) + "  Links: " + contact.link_names[0] + ", " +
+                                   contact.link_names[1] + "  Distance: " + std::to_string(contact.distance))
+                                      .c_str());
+    throw std::runtime_error("In OMPLDefaultPlanProfile: All goal states are either in collision or outside limits");
+  }
+
+  return valid_solutions;
+}
```

Comment:
It takes a surprisingly large amount of code to perform collision-aware IK. We should consider making this a capability of the `KinematicGroup` or a utility function somewhere so we don't have to continually rewrite this

Reply:
Is the same code used anywhere else? I believe this is the only place where this is being done. 

For others that may read this it is doing more than just collision-aware IK. It checks and enforces bounds, calculates redundant solutions, along with print debug information if no valid solution is found.

Reply:
It's not necessarily being used elsewhere yet, but having this collision-aware IK capability is really valuable to a lot of other applications. In those cases, someone would have to reimplement this function because it doesn't formally exist in `tesseract`. My suggestion would be to move this somewhere like the `KinematicsGroup` to expose it to other applications because it is generic and valuable

Reply:
I agree, the only thing with adding it to the kinematic group is it would now make tesseract_kinematics depend on tesseract_collision which I am not opposed to but wanted to mention it.

Reply:
Yeah, I thought about that as well. Would it make sense to move the high-level manipulator classes (`KinematicGroup` and possibly `JointGroup` for consistency) to `tesseract_environment` where they could have access to collision checkers? In my mind the value of the manipulator classes is that they unite the various underlying pieces (IK, FK, enforcing joint limits, getting redundant solutions, checking collision, etc.) for moving a robot into a high-level easily usable class. `tesseract_environment` seems to be the container for that high-level capability